namecoin/src/main/java/org/libdohj/names/NameLookupLatestLevelDBTransactionCache.java

```diff
@@ -0,0 +1,410 @@
+        // height goes here
+        byte[] txBytes = tx.bitcoinSerialize();
+        
+        ByteBuffer keyBuffer = ByteBuffer.allocate(headerBytes.length + nameBytes.length);
+        ByteBuffer recordBuffer = ByteBuffer.allocate(4 + txBytes.length);
+        
+        keyBuffer.put(headerBytes).put(nameBytes);
+        recordBuffer.putInt(height).put(txBytes);
+        
+        db.put(keyBuffer.array(), recordBuffer.array());
+    }
+    
+    // TODO: stop duplicating code from the other NameLookupLatest implementations
+    protected void verifyHeightTrustworthy(int height) throws Exception {
+        if (height < 1) {
+            throw new Exception("Nonpositive block height; not trustworthy!");
+        }
+        
+        int headHeight = chain.getChainHead().getHeight();
+        
+        int confirmations = headHeight - height + 1;
+        
+        // TODO: optionally use transaction chains (with signature checks) to verify transactions without 12 confirmations
+        // TODO: the above needs to be optional, because some applications (e.g. cert transparency) require confirmations
+        if (confirmations < 12) {
+            throw new Exception("Block does not yet have 12 confirmations; not trustworthy!");
+        }
+        
+        // TODO: check for off-by-one errors on this line
+        if (confirmations >= 36000) {
+            throw new Exception("Block has expired; not trustworthy!");
+        }
+    }
+    
+    // TODO: make a new Exception class
+    @Override
+    public Transaction getNameTransaction(String name, String identity) throws Exception {
+        
+        byte[] headerBytes = "NameTx".getBytes("ISO-8859-1");
+        byte[] nameBytes = name.getBytes("ISO-8859-1");
+        // name goes here
+        
+        ByteBuffer keyBuffer = ByteBuffer.allocate(headerBytes.length + nameBytes.length);
+        keyBuffer.put(headerBytes).put(nameBytes);
+        
+        byte[] recordBytes = db.get(keyBuffer.array());
+        if (recordBytes == null)
+            return null;
+        
+        ByteBuffer recordBuffer = ByteBuffer.wrap(recordBytes);
+        
+        int height = recordBuffer.getInt();
+        
+        verifyHeightTrustworthy(height);
+        
+        Transaction tx = new Transaction(params, recordBytes, 4);
+        
+        tx.getConfidence().setAppearedAtChainHeight(height); // TODO: test this line
+        tx.getConfidence().setDepthInBlocks(chain.getChainHead().getHeight() - height + 1);
+        
+        return tx;
+    }
+    
+    protected synchronized int getChainHead() {
+        return ByteBuffer.wrap(db.get(CHAIN_HEAD_KEY)).getInt();
+    }
+    
+    protected synchronized void setChainHead(int chainHead) {
+        db.put(CHAIN_HEAD_KEY, ByteBuffer.allocate(4).putInt(chainHead).array());
+    }
+    
+    public synchronized void close() throws Exception {
+        try {
+            db.close();
+        } catch (IOException e) {
+            throw new Exception(e);
```

Comment:
Why catch then wrap and rethrow? Surely this should just be;

`public synchronized void close() throws IOException {`
`    db.close();`
`}`